File: tdf-hq-ui/src/api/domoQuotes.ts
import { get, post } from './client';
import type { components } from './generated/types';
import type { DatafastCheckoutDTO, PaypalCreateDTO } from './types';

export type PublicDomoStorefront = components['schemas']['PublicDomoStorefront'];
export type PublicDomoQuoteCreateRequest = components['schemas']['PublicDomoQuoteCreateRequest'];
export type PublicDomoQuote = components['schemas']['PublicDomoQuote'];

const quoteBase = (quoteId: string): string => {
  const normalized = quoteId.trim();
  if (!/^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(normalized)) {
    throw new Error('quoteId must be a UUID.');
  }
  return `/public/domo/quotes/${normalized}`;
};

const lookupHeaders = (lookupToken: string): RequestInit => ({
  headers: { 'X-Order-Lookup-Token': lookupToken },
});

export const DomoQuotes = {
  getStorefront: () => get<PublicDomoStorefront>('/public/domo'),
  createQuote: (payload: PublicDomoQuoteCreateRequest, idempotencyKey: string) =>
    post<PublicDomoQuote>('/public/domo/quotes', payload, {
      headers: { 'Idempotency-Key': idempotencyKey },
    }),
  getQuote: (quoteId: string, lookupToken: string) =>
    get<PublicDomoQuote>(quoteBase(quoteId), lookupHeaders(lookupToken)),
  acceptQuote: (quoteId: string, lookupToken: string) =>
    post<PublicDomoQuote>(
      `${quoteBase(quoteId)}/accept`,
      { termsAccepted: true },
      lookupHeaders(lookupToken),
    ),
  createDatafastCheckout: (quoteId: string, lookupToken: string) =>
    post<DatafastCheckoutDTO>(
      `${quoteBase(quoteId)}/datafast/checkout`,
      {},
      lookupHeaders(lookupToken),
    ),
  confirmDatafastStatus: (quoteId: string, resourcePath: string, lookupToken: string) => {
    const search = new URLSearchParams({ resourcePath });
    return get<PublicDomoQuote>(
      `${quoteBase(quoteId)}/datafast/status?${search.toString()}`,
      lookupHeaders(lookupToken),
    );
  },
  createPaypalOrder: (quoteId: string, lookupToken: string) =>
    post<PaypalCreateDTO>(
      `${quoteBase(quoteId)}/paypal/create`,
      {},
      lookupHeaders(lookupToken),
    ),
  capturePaypalOrder: (
    quoteId: string,
    paypalOrderId: string,
    lookupToken: string,
  ) => post<PublicDomoQuote>(
    `${quoteBase(quoteId)}/paypal/capture`,
    { paypalOrderId },
    lookupHeaders(lookupToken),
  ),
};

File: tdf-hq/src/TDF/Routes/DomoQuotes.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}

module TDF.Routes.DomoQuotes
  ( PublicDomoStorefrontDTO(..)
  , PublicDomoQuoteCreateRequest(..)
  , PublicDomoQuoteAcceptRequest(..)
  , PublicDomoQuoteLineDTO(..)
  , PublicDomoQuoteDTO(..)
  , PublicDomoPaypalCaptureRequest(..)
  , PublicDomoQuotesAPI
  ) where

import           Data.Aeson
  ( FromJSON(..), Options(..), ToJSON, defaultOptions, genericParseJSON )
import           Data.Int (Int64)
import           Data.Text (Text)
import           Data.Time (UTCTime)
import           GHC.Generics (Generic)
import           Servant

import qualified TDF.API.Types as APITypes
import           TDF.API.Types (rejectNullOptionalFields)

data PublicDomoStorefrontDTO = PublicDomoStorefrontDTO
  { checkoutAvailable    :: Bool
  , unavailableReason    :: Maybe Text
  , rateCardVersion      :: Maybe Text
  , currency             :: Maybe Text
  , eventTypes           :: [Text]
  , maximumGuests        :: Maybe Int
  , maximumDurationHours :: Maybe Int
  , maximumSetupHours    :: Maybe Int
  , quoteHoldMinutes     :: Maybe Int
  , timezone             :: Text
  } deriving (Eq, Show, Generic)

instance ToJSON PublicDomoStorefrontDTO

data PublicDomoQuoteCreateRequest = PublicDomoQuoteCreateRequest
  { customerName  :: Text
  , customerEmail :: Text
  , customerPhone :: Maybe Text
  , eventType     :: Text
  , guests        :: Int
  , startsAt      :: UTCTime
  , durationHours :: Int
  , setupHours    :: Int
  , catering      :: Bool
  , production    :: Bool
  , transport     :: Bool
  , notes         :: Maybe Text
  } deriving (Eq, Show, Generic)

instance ToJSON PublicDomoQuoteCreateRequest
instance FromJSON PublicDomoQuoteCreateRequest where
  parseJSON value = do
    rejectNullOptionalFields
      "PublicDomoQuoteCreateRequest"
      ["customerPhone", "notes"]
      value
    genericParseJSON defaultOptions { rejectUnknownFields = True } value

data PublicDomoQuoteAcceptRequest = PublicDomoQuoteAcceptRequest
  { termsAccepted :: Bool
  } deriving (Eq, Show, Generic)

instance ToJSON PublicDomoQuoteAcceptRequest
instance FromJSON PublicDomoQuoteAcceptRequest where
  parseJSON = genericParseJSON defaultOptions { rejectUnknownFields = True }

data PublicDomoQuoteLineDTO = PublicDomoQuoteLineDTO
  { code            :: Text
  , description     :: Text
  , quantity        :: Int
  , unitAmountMinor :: Int64
  , subtotalMinor   :: Int64
  } deriving (Eq, Show, Generic)

instance ToJSON PublicDomoQuoteLineDTO

data PublicDomoQuoteDTO = PublicDomoQuoteDTO
  { quoteId           :: Text
  , checkoutId        :: Text
  , lookupToken       :: Maybe Text
  , quoteStatus       :: Text
  , paymentStatus     :: Text
  , fulfillmentStatus :: Text
  , rateCardVersion   :: Text
  , currency          :: Text
  , eventType         :: Text
  , guests            :: Int
  , startsAt          :: UTCTime
  , endsAt            :: UTCTime
  , setupStartsAt     :: UTCTime
  , lines             :: [PublicDomoQuoteLineDTO]
  , subtotalMinor     :: Int64
  , taxMinor          :: Int64
  , totalMinor        :: Int64
  , depositMinor      :: Int64
  , balanceMinor      :: Int64
  , timezone          :: Text
  , termsVersion      :: Text
  , holdExpiresAt     :: UTCTime
  , termsAcceptedAt   :: Maybe UTCTime
  , depositPaidAt     :: Maybe UTCTime
  , paymentMethods    :: [Text]
  } deriving (Eq, Show, Generic)

instance ToJSON PublicDomoQuoteDTO

data PublicDomoPaypalCaptureRequest = PublicDomoPaypalCaptureRequest
  { paypalOrderId :: Text
  } deriving (Eq, Show, Generic)

instance ToJSON PublicDomoPaypalCaptureRequest
instance FromJSON PublicDomoPaypalCaptureRequest where
  parseJSON = genericParseJSON defaultOptions { rejectUnknownFields = True }

type PublicDomoQuotesAPI =
       "public" :> "domo" :> Get '[JSON] PublicDomoStorefrontDTO
  :<|> "public" :> "domo" :> "quotes"
         :> Header "Idempotency-Key" Text
         :> ReqBody '[JSON] PublicDomoQuoteCreateRequest
         :> PostCreated '[JSON] PublicDomoQuoteDTO
  :<|> "public" :> "domo" :> "quotes" :> Capture "quoteId" Text
         :> Header "X-Order-Lookup-Token" Text
         :> Get '[JSON] PublicDomoQuoteDTO
  :<|> "public" :> "domo" :> "quotes" :> Capture "quoteId" Text :> "accept"
         :> Header "X-Order-Lookup-Token" Text
         :> ReqBody '[JSON] PublicDomoQuoteAcceptRequest
         :> Post '[JSON] PublicDomoQuoteDTO
  :<|> "public" :> "domo" :> "quotes" :> Capture "quoteId" Text
         :> "datafast" :> "checkout"
         :> Header "X-Order-Lookup-Token" Text
         :> Post '[JSON] APITypes.DatafastCheckoutDTO
  :<|> "public" :> "domo" :> "quotes" :> Capture "quoteId" Text
         :> "datafast" :> "status"
         :> Header "X-Order-Lookup-Token" Text
         :> QueryParam' '[Required] "resourcePath" Text
         :> Get '[JSON] PublicDomoQuoteDTO
  :<|> "public" :> "domo" :> "quotes" :> Capture "quoteId" Text
         :> "paypal" :> "create"
         :> Header "X-Order-Lookup-Token" Text
         :> Post '[JSON] APITypes.PaypalCreateDTO
  :<|> "public" :> "domo" :> "quotes" :> Capture "quoteId" Text
         :> "paypal" :> "capture"
         :> Header "X-Order-Lookup-Token" Text
         :> ReqBody '[JSON] PublicDomoPaypalCaptureRequest
         :> Post '[JSON] PublicDomoQuoteDTO
